File: internal/service/permission.go
package service

import (
	"context"

	"github.com/ciclebyte/template_starter/api/v1/permission"
)

// ============================================================================
// 权限管理服务接口
// ============================================================================

type (
	IPermission interface {
		// 权限管理
		ListPermissions(ctx context.Context, req *permission.ListPermissionsReq) (*permission.ListPermissionsRes, error)
		CreatePermission(ctx context.Context, req *permission.CreatePermissionReq) (*permission.CreatePermissionRes, error)
		UpdatePermission(ctx context.Context, req *permission.UpdatePermissionReq) (*permission.UpdatePermissionRes, error)
		DeletePermission(ctx context.Context, req *permission.DeletePermissionReq) (*permission.DeletePermissionRes, error)
		GetPermission(ctx context.Context, req *permission.GetPermissionReq) (*permission.GetPermissionRes, error)

		// 角色管理
		ListRoles(ctx context.Context, req *permission.ListRolesReq) (*permission.ListRolesRes, error)
		CreateRole(ctx context.Context, req *permission.CreateRoleReq) (*permission.CreateRoleRes, error)
		UpdateRole(ctx context.Context, req *permission.UpdateRoleReq) (*permission.UpdateRoleRes, error)
		DeleteRole(ctx context.Context, req *permission.DeleteRoleReq) (*permission.DeleteRoleRes, error)
		GetRole(ctx context.Context, req *permission.GetRoleReq) (*permission.GetRoleRes, error)
		AssignRolePermissions(ctx context.Context, req *permission.AssignRolePermissionsReq) (*permission.AssignRolePermissionsRes, error)

		// 用户角色管理
		ListUserRoles(ctx context.Context, req *permission.ListUserRolesReq) (*permission.ListUserRolesRes, error)
		AssignUserRoles(ctx context.Context, req *permission.AssignUserRolesReq) (*permission.AssignUserRolesRes, error)
		RemoveUserRole(ctx context.Context, req *permission.RemoveUserRoleReq) (*permission.RemoveUserRoleRes, error)
	}
)

var (
	localPermission IPermission
)

func Permission() IPermission {
	if localPermission == nil {
		panic("implement not found for interface IPermission, forgot register?")
	}
	return localPermission
}

func RegisterPermission(i IPermission) {
	localPermission = i
}

File: web/src/api/permission.js
import request from '@/utils/request'

// ============================================================================
// 权限管理
// ============================================================================

// 获取权限列表
export function getPermissions(params) {
  return request({
    url: '/api/v1/permissions',
    method: 'get',
    params
  })
}

// 创建权限
export function createPermission(data) {
  return request({
    url: '/api/v1/permissions',
    method: 'post',
    data
  })
}

// 更新权限
export function updatePermission(id, data) {
  return request({
    url: `/api/v1/permissions/${id}`,
    method: 'put',
    data
  })
}

// 删除权限
export function deletePermission(id) {
  return request({
    url: `/api/v1/permissions/${id}`,
    method: 'delete'
  })
}

// 获取权限详情
export function getPermission(id) {
  return request({
    url: `/api/v1/permissions/${id}`,
    method: 'get'
  })
}

// ============================================================================
// 角色管理
// ============================================================================

// 获取角色列表
export function getRoles(params) {
  return request({
    url: '/api/v1/roles',
    method: 'get',
    params
  })
}

// 创建角色
export function createRole(data) {
  return request({
    url: '/api/v1/roles',
    method: 'post',
    data
  })
}

// 更新角色
export function updateRole(id, data) {
  return request({
    url: `/api/v1/roles/${id}`,
    method: 'put',
    data
  })
}

// 删除角色
export function deleteRole(id) {
  return request({
    url: `/api/v1/roles/${id}`,
    method: 'delete'
  })
}

// 获取角色详情
export function getRole(id) {
  return request({
    url: `/api/v1/roles/${id}`,
    method: 'get'
  })
}

// 分配角色权限
export function assignRolePermissions(id, data) {
  return request({
    url: `/api/v1/roles/${id}/permissions`,
    method: 'post',
    data
  })
}

// ============================================================================
// 用户角色管理
// ============================================================================

// 获取用户角色列表
export function getUserRoles(userId) {
  return request({
    url: `/api/v1/users/${userId}/roles`,
    method: 'get'
  })
}

// 分配用户角色
export function assignUserRoles(userId, data) {
  return request({
    url: `/api/v1/users/${userId}/roles`,
    method: 'post',
    data
  })
}

// 移除用户角色
export function removeUserRole(userId, roleId) {
  return request({
    url: `/api/v1/users/${userId}/roles/${roleId}`,
    method: 'delete'
  })
}